File: 1669. Merge In Between Linked Lists/mergeInBetween.js
// Definition for singly-linked list.
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

// helper function to generate LinkedList from array
const arrayToLinkedList = (arr) => {
  if (!arr || !arr.length) {
    return null;
  }

  const dummyHead = new ListNode(0);
  let curr = dummyHead;

  for (let i = 0; i < arr.length; i++) {
    let node = new ListNode(arr[i]);
    curr.next = node;
    curr = curr.next;
  }

  return dummyHead.next;
}

/**
 * @param {ListNode} list1
 * @param {number} a
 * @param {number} b
 * @param {ListNode} list2
 * @return {ListNode}
 */
const mergeInBetween = (list1, a, b, list2) => {
  let list1AthNode = list1;
  let list1BthNode = list1;

  a--;
  while (a > 0) {
    list1AthNode = list1AthNode.next;
    a--;
  }

  b++;
  while (b > 0) {
    list1BthNode = list1BthNode.next;
    b--;
  }

  let list2LastNode = list2
  while (list2LastNode.next) {
    list2LastNode = list2LastNode.next
  }

  list1AthNode.next = list2
  list2LastNode.next = list1BthNode

  return list1
};

const list1 = arrayToLinkedList([10, 1, 13, 6, 9, 5]);
const list2 = arrayToLinkedList([1000000, 1000001, 1000002]);
// [10,1,13,1000000,1000001,1000002,5]
const a = 3;
const b = 4;
console.log('mergeInBetween: ', mergeInBetween(list1, a, b, list2).next)

File: 1669. Merge In Between Linked Lists/mergeInBetween.py
from typing import Optional, List
# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next

# helper function to generate LinkedList from array
def arrayToLinkedList(arr: Optional[List]) -> Optional[ListNode]:
  if not arr:
    return None

  dummyHead = ListNode(0)
  curr = dummyHead

  for i in range(len(arr)):
    node = ListNode(arr[i])
    curr.next = node
    curr = curr.next
  
  return dummyHead.next

class Solution:
  def mergeInBetween(self, list1: ListNode, a: int, b: int, list2: ListNode) -> ListNode:
    list1AthNode = list1
    list1BthNode = list1
  
    a -= 1
    while a > 0:
      list1AthNode = list1AthNode.next
      a -= 1

    b += 1
    while b > 0:
      list1BthNode = list1BthNode.next
      b -= 1

    list2LastNode = list2
    while list2LastNode.next:
      list2LastNode = list2LastNode.next

    list1AthNode.next = list2
    list2LastNode.next = list1BthNode

    return list1


list1 = arrayToLinkedList([10, 1, 13, 6, 9, 5])
list2 = arrayToLinkedList([1000000, 1000001, 1000002])
# [10,1,13,1000000,1000001,1000002,5]
a = 3
b = 4
solution = Solution()
print('mergeInBetween: ', solution.mergeInBetween(list1, a, b, list2))
